pkg/execution/state/redis_state/queue_processor.go

```diff
@@ -498,30 +509,124 @@ func (q *queue) worker(ctx context.Context, f osqueue.RunFunc) {
 // shadowWorker runs a blocking process that listens to item being pushed into the
 // shadow queue partition channel. This allows us to process an individual shadow partition.
 // TODO: replace channel type with QueueShadowPartition struct once available
-func (q *queue) shadowWorker(ctx context.Context, qspc chan *int) {
+func (q *queue) shadowWorker(ctx context.Context, qspc chan *QueueShadowPartition) {
 	for {
 		select {
 		case <-ctx.Done():
 			return
 
-		case <-qspc:
-			metrics.ActiveShadowScanerCount(ctx, 1, metrics.CounterOpt{PkgName: pkgName})
+		case shadowPart := <-qspc:
+			err := q.processShadowPartition(ctx, shadowPart, 0)
+			if err != nil {
+				logger.StdlibLogger(ctx).Error("could not scan shadow partition", "err", err, "shadow_part", shadowPart)
+			}
+		}
+	}
+}
+
+func (q *queue) processShadowPartition(ctx context.Context, shadowPart *QueueShadowPartition, continuationCount uint) error {
+	metrics.ActiveShadowScannerCount(ctx, 1, metrics.CounterOpt{PkgName: pkgName})
+	defer metrics.ActiveShadowScannerCount(ctx, -1, metrics.CounterOpt{PkgName: pkgName})
 
-			// TODO:
-			//
-			// - retrieve countinuation counter
-			//
-			// loop:
-			// - extend lease for shadow partition
-			// - iterate through backlogs
-			//   + add item to function partition
-			//
+	// acquire lease for shadow partition
+	leaseID, err := duration(ctx, q.primaryQueueShard.Name, "shadow_partition_lease", q.clock.Now(), func(ctx context.Context) (*ulid.ULID, error) {
+		leaseID, err := q.ShadowPartitionLease(ctx, shadowPart, ShadowPartitionLeaseDuration)
+		return leaseID, err
+	})
+	if err != nil {
+		if errors.Is(err, ErrShadowPartitionAlreadyLeased) {
+			// contention
+			return nil
+		}
+	}
+
+	if leaseID == nil {
+		return fmt.Errorf("missing shadow partition leaseID")
+	}
+
+	extendLeaseCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// keep extending lease while we're processing
+	go func() {
+		for {
+			select {
+			case <-extendLeaseCtx.Done():
+				return
+			case <-time.Tick(ShadowPartitionLeaseDuration / 2):
+				leaseID, err = q.ShadowPartitionExtendLease(ctx, shadowPart, *leaseID, ShadowPartitionLeaseDuration)
+				if err != nil {
+					if errors.Is(err, ErrShadowPartitionAlreadyLeased) || errors.Is(err, ErrShadowPartitionLeaseNotFound) {
+						// contention
+						return
+					}
+					return
+				}
+			}
+		}
+	}()
+
+	// Check if shadow partition cannot be processed (paused/refill disabled, etc.)
+	if shadowPart.PauseRefill {
+		q.removeShadowContinue(ctx, shadowPart, false)
+
+		forceRequeueAt := q.clock.Now().Add(ShadowPartitionRefillPausedRequeueExtension)
+		err = q.ShadowPartitionRequeue(ctx, shadowPart, *leaseID, &forceRequeueAt)
```

Comment:
do we actually need to requeue?
seems like a redundant work that only consume additional qps for no benefit.

besides when unpaused, it'll just start processing from the oldest stuff it sees right away.

Reply:
right, but why the need to do so?

Reply:
Because we don't want to shadow scan paused functions?